### PR TITLE
fix(linker): Fix yarn removing linked deps during link stage

### LINKED ISSUE #3288: yarn overwrites symlinked dependencies
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**
I have a development checkout of the package `@company/foo` which I made globally available through `yarn link`. I then linked that packages into my project `@company/project`.

I also have package `@company/bar`, which is linked identically into `@company/project`.

When I now run `yarn upgrade @company/bar` in the directory of `@company/project`, yarn will overwrite my local development checkout with data it pulled from the npm registry.

**If the current behavior is a bug, please provide the steps to reproduce.**

It should be reproducible with any combination of linked packages as outlined above. It seems to me that when yarn says "linking dependencies", it will simply pull some fixed version from the local cache and overwrite my local development snapshot.

**What is the expected behavior?**
The linked package is not overwritten. npm deals with this: https://github.com/npm/npm/commit/1619871ac0cc8839dc9962c78e736095976c1eb4

**Please mention your node.js, yarn and operating system version.**
6.10.2, 0.23.3, Windows 10

### LINKED ISSUE #4603: `yarn install`1.1.0 overwrites symlinks created by `yarn link`
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
When using `yarn link <local-package>` a symlink is created in `node_modules/`. After running `yarn install`, the symlink is removed with the public registry version of the local module.

**If the current behavior is a bug, please provide the steps to reproduce.**
https://github.com/javinor/yarn-removes-symlinks using yarn 1.1.0

**What is the expected behavior?**
symlinks should not be removed when running `yarn install` as happens in yarn v0.27.5

**Please mention your node.js, yarn and operating system version.**
macOS Sierra
node 6.9.2
yarn 1.1.0 (doesn't happen with yarn 0.27.5)

### LINKED ISSUE #4635: `yarn add` is losing workspace symlinks
Using yarn 1.1.0, node 8.3.0:

```sh
ls node_modules # includes several symlinks from workspace modules
yarn add --dev jest
ls node_modules # symlinks are gone
```

There are no errors during the `yarn add`, it just seems to blow away workspace links.  Running `yarn` again restores them.

### LINKED ISSUE #4770: Running yarn link unlinks any modules linked via yarn link.
Yarn version 1.2.1.

**Do you want to request a *feature* or report a *bug*?**

Bug.

**What is the current behavior?**

Running `yarn add [package-name]` unlinks any linked modules. Here is a screenshot showing this:

screen shot 2017-10-24 at 10 18 15 am

**If the current behavior is a bug, please provide the steps to reproduce.**

`cd app`
`ls -l node_modules | grep sequential` --> shows that the module is not linked
`yarn link sequential-ids`
`ls -l node_modules | grep sequential` --> shows that the module is now linked
`yarn add lodash`
`ls -l node_modules | grep sequential` --> shows that the module is now unlinked

**What is the expected behavior?**

`yarn add [package-name]` should never cause linked packages to become unlinked.

**Please mention your node.js, yarn and operating system version.**

Node 8.6.0.

macOS High Sierra. This also happened on Sierra.

### PR DESCRIPTION
**Pull request by gandazgul**
**Summary**

 Actual fix: changed fs.readlink to fs.realpath when checking if a symlink is a linked dependency in package-linker.js This fixes yarn removing linked deps when installing or updating.

Fixes #3288, fixes #4770, fixes #4635, fixes #4603.

Potential fix for #3202.

**Test plan**

See https://github.com/yarnpkg/yarn/issues/3288#issuecomment-335955366 for repro steps.
See https://github.com/yarnpkg/yarn/issues/3288#issuecomment-338503103 for my explanation of the problem.

With a real world test scenario this works, but I'm unable to have it break from a unit test. I added a test in the integration suite but with the bug added back in it still passes because both generated paths are identical. I would like some help with the unit test.

### CODE REVIEW

**Comment by gandazgul** [comment]
Note to myself, found a bug, yarn throws when a linked module's destination no longer exists, try catch?

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "yarn",
   "installationMethod": "unknown",
-  "version": "1.2.1",
+  "version": "1.2.2",
```
Comment (bestander) [inline package.json]: this is unnecessary, the core team will bump the patch version when they are ready to release it

File: __tests__/fixtures/install/install-dont-overwrite-linked/.npmrc [hunk 2]
```diff
@@ -0,0 +1 @@
+yarn-offline-mirror=./mirror-for-offline
```
Comment (bestander) [inline __tests__/fixtures/install/install-dont-overwrite-linked/.npmrc]: this file is not needed for the test

File: src/package-linker.js [hunk 3]
```diff
@@ -304,8 +304,15 @@ export default class PackageLinker {
       const stat = await fs.lstat(entryPath);
 
       if (stat.isSymbolicLink()) {
-        const packageName = entry;
-        linkTargets.set(packageName, await fs.readlink(entryPath));
+        try {
```
Comment (bestander) [inline src/package-linker.js]: eslint errors on this.
Please run `yarn run prettier` to autoformat the code

File: __tests__/fixtures/install/install-dont-overwrite-linked/.npmrc (hunk 2 above)
Comment (bestander) [reply __tests__/fixtures/install/install-dont-overwrite-linked/.npmrc]: nevermind, I was wrong here

File: __tests__/commands/install/integration.js [hunk 4]
```diff
@@ -865,6 +865,29 @@ test.concurrent('install will not overwrite files in symlinked scoped directorie
   );
 });
 
+test.concurrent('install will not overwrite linked dependencies', async (): Promise<void> => {
+  await runInstall(
+    {},
+    'install-dont-overwrite-linked',
+    async (config): Promise<void> => {
+      const dependencyPath = path.join(config.cwd, 'node_modules', 'fake-dependency');
```
Comment (bestander) [inline __tests__/commands/install/integration.js]: I don't understand the test.
i would expect that you would run a link command and then install should not overwrite it?

Comment (bestander) [reply __tests__/commands/install/integration.js]: A few tips.

0. You probably want to setup several folders for the test: a linked package.json and the linker package.json
1. here is how link command is used in tests https://github.com/yarnpkg/yarn/blob/master/__tests__/commands/link.js#L35
2. after linking you can check that link exists
3. after that run install command and check that link was not destroyed
https://github.com/yarnpkg/yarn/blob/master/__tests__/commands/install/integration.js#L93

File: package.json (hunk 1 above)
Comment (gandazgul) [reply package.json]: I will revert this. Thank you.

File: src/package-linker.js (hunk 3 above)
Comment (gandazgul) [reply src/package-linker.js]: Will do, and will check with the linter before committing. Thanks.

File: __tests__/commands/install/integration.js (hunk 4 above)
Comment (gandazgul) [reply __tests__/commands/install/integration.js]: Thank you! Yeah I dont get this test either. I copied it from the other one that checks if scoped links are maintained, I suspect that one is faulty too. 

I'll update both with your suggestions here.

File: src/package-linker.js (hunk 3 above)
Comment (gandazgul) [reply src/package-linker.js]: Done. yarn lint completed.

**Comment by gandazgul** [comment]
I re-wrote the tests with @bestander suggestions and tested them with the old code to make sure they break. Both tests for scoped and regular packages now fail correctly and pass correctly with the new code.

File: __tests__/commands/_helpers.js [hunk 5]
```diff
@@ -113,11 +125,16 @@ export async function run<T, R>(
...
     const source = typeof name === 'string' ? name : name.source;
 
-    const dir = path.join(fixturesLoc, source);
-    cwd = await fs.makeTempDir(path.basename(dir));
-    await fs.copy(dir, cwd, reporter);
-    if (typeof name !== 'string') {
-      cwd = path.join(cwd, name.cwd);
+    // if source wasn't set then assume we were given a complete path
+    if (typeof source === 'undefined') {
+      cwd = typeof name !== 'string' ? name.cwd : await fs.makeTempDir();
```
Comment (BYK) [inline __tests__/commands/_helpers.js]: These nested conditionals are a bit confusing. Do you think we can find a cleaner way to do this?

File: __tests__/commands/link.js [hunk 6]
```diff
@@ -1,24 +1,12 @@
 /* @flow */
 
-import {run as buildRun} from './_helpers.js';
-import {run as link} from '../../src/cli/commands/link.js';
+import {runLink} from './_helpers.js';
```
Comment (BYK) [inline __tests__/commands/link.js]: Nice refactor.

File: src/package-linker.js [hunk 7]
```diff
@@ -317,7 +322,13 @@ export default class PackageLinker {
 
           if (stat2.isSymbolicLink()) {
             const packageName = `${scopeName}/${entry2}`;
-            linkTargets.set(packageName, await fs.readlink(entryPath2));
+            try {
```
Comment (BYK) [inline src/package-linker.js]: This repeated code kind of calls for a helper function.

**Comment by BYK** [comment]
LGTM. @bestander approve and merge at will!

**Comment by BYK** [comment]
>Fixes: #3288 and other reported issues relating to linked deps being removed.

Do we have a list for these?

File: src/package-linker.js (hunk 7 above)
Comment (gandazgul) [reply src/package-linker.js]: Yeah I was thinking about it. But I didn't want to confuse this too much. There's also a todo comment to merge the link targets logic here with logic in another part of the code. I can take a look after and make another PR.

**Comment by gandazgul** [comment]
@BYK I found these: 

https://github.com/yarnpkg/yarn/issues/4770
https://github.com/yarnpkg/yarn/issues/4635
https://github.com/yarnpkg/yarn/issues/4603

File: __tests__/commands/_helpers.js (hunk 5 above)
Comment (gandazgul) [reply __tests__/commands/_helpers.js]: they are, if I take this one off which is unnecessary then flow complains. I will take a look in another PR see if I can can come up with a clearer logic and put more comments.

**Comment by bestander** [comment]
Great job, @gandazgul!
Thanks for helping with review, @BYK

**Comment by bertho-zero** [comment]
It works perfectly, npm has this problem since the beginning of version 3, thank you very much!

**Comment by gandazgul** [comment]
@bestander thank you, happy to help. When do you think this will be released? My team is anxious to upgrade yarn to the latest version.

**Comment by bestander** [comment]
@gandazgul it is due to be released any day soon according to recent release history.
In the mean time feel free to use nightly builds https://yarnpkg.com/en/docs/nightly

**Comment by rdsedmundo** [comment]
This PR hasn't fixed this problem entirely. I'm still facing this issue.

NodeJS version: `10.0.0`
Yarn version: `1.6.0`

I do the `yarn link package` and I can confirm it works by executing `ls -l node_modules/package` and seeing the symlink pointing to the right folder. After I run `yarn install` though and execute the `ls -l node_modules/package` command again, the symlink is removed and it does use the package from what is on `package.json` again (it's a Git URL by the way, don't know if it makes difference).

**Pull request merged**